openapi/controller/main.py

```diff
@@ -0,0 +1,183 @@
+# -*- coding: utf-8 -*-
+import functools
+import hashlib
+import logging
+import os
+from ast import literal_eval
+import base64
+
+import odoo
+from odoo import http, SUPERUSER_ID, models
+from odoo.http import request, OpenERPSession
+from odoo.modules.registry import RegistryManager
+from psycopg2.extensions import ISOLATION_LEVEL_READ_COMMITTED
+
+from .pinguin import *
+
+_logger = logging.getLogger(__name__)
+
+#################################################################
+# Odoo REST API                                                 #
+#  Version 1                                                    #
+# --------------------------------------------------------------#
+# The current api version is considered stable, although        #
+# the exposed models and methods change as they are configured  #
+# on the database level. Only if significant changes in the api #
+# generation logic should be implemented in the future          #
+# a version bump should be considered.                          #
+#################################################################
+
+API_ENDPOINT = '/api'
+API_ENDPOINT_V1 = '/v1'
+# API_ENDPOINT_V2 = '/2'
+
+# We patch the route decorator in pinguin.py
+# with authentication and DB inference logic.
+# We also check if the model is installed in the database.
+# This keeps the code below minial and readable.
+http.route = pinguin_route
+
+
+class ApiV1Controller(http.Controller):
+    """ Implements the REST API V1 endpoint.
+    .. methods:
+
+        CRUD Methods:
+        - `POST     ../<model>`               -> `CreateOne`
```

Comment:
two dots looks like moving to upper folder. Use three dots, without any prefix or full address::

    .../<model>
    <model>
    /api/<version>/<model>